src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -987,41 +990,58 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
         }
     }
 
-    private function getValidatorMappingFiles(ContainerBuilder $container)
+    private function getValidatorMappingFiles(ContainerBuilder $container, array &$files)
     {
-        $files = array(array(), array());
-
         if (interface_exists('Symfony\Component\Form\FormInterface')) {
             $reflClass = new \ReflectionClass('Symfony\Component\Form\FormInterface');
-            $files[0][] = dirname($reflClass->getFileName()).'/Resources/config/validation.xml';
-            $container->addResource(new FileResource($files[0][0]));
+            $files['xml'][] = $file = dirname($reflClass->getFileName()).'/Resources/config/validation.xml';
+            $container->addResource(new FileResource($file));
         }
 
         foreach ($container->getParameter('kernel.bundles_metadata') as $bundle) {
             $dirname = $bundle['path'];
-            if (is_file($file = $dirname.'/Resources/config/validation.xml')) {
-                $files[0][] = $file;
+
+            if (is_file($file = $dirname.'/Resources/config/validation.yml')) {
+                $files['yml'][] = $file;
                 $container->addResource(new FileResource($file));
             }
 
-            if (is_file($file = $dirname.'/Resources/config/validation.yml')) {
-                $files[1][] = $file;
+            if (is_file($file = $dirname.'/Resources/config/validation.xml')) {
+                $files['xml'][] = $file;
                 $container->addResource(new FileResource($file));
             }
 
             if (is_dir($dir = $dirname.'/Resources/config/validation')) {
-                foreach (Finder::create()->followLinks()->files()->in($dir)->name('*.xml') as $file) {
-                    $files[0][] = $file->getPathname();
-                }
-                foreach (Finder::create()->followLinks()->files()->in($dir)->name('*.yml') as $file) {
-                    $files[1][] = $file->getPathname();
-                }
-
+                $this->getValidatorMappingFilesFromDir($dir, $files);
                 $container->addResource(new DirectoryResource($dir));
             }
         }
+    }
+
+    private function getValidatorMappingFilesFromDir($dir, array &$files)
+    {
+        foreach (Finder::create()->followLinks()->files()->in($dir)->name('/\.(xml|ya?ml)$/') as $file) {
+            $extension = $file->getExtension();
+            $files['yaml' === $extension ? 'yml' : $extension][] = $file->getRealpath();
+        }
+    }
 
-        return $files;
+    private function getValidatorMappingFilesFromConfig(array $config, array &$files)
+    {
+        foreach ($config['mapping']['paths'] as $path) {
+            if (is_dir($path)) {
+                $this->getValidatorMappingFilesFromDir($path, $files);
+            } elseif (is_file($path)) {
```

Comment:
this must use the ContainerBuilder API to register resources to invalidate the cache when files are changing in debug mode